README.md
# cluster-info-client

A JavaScript API for securely fetching information about groups of clusters.

It stores all cluster information in Vault.

It also manages grouping clusters into "channels". Channels are free-form, and a cluster can be placed into any number of channels.  Example uses for channels are deployment groups, environments, zones, or any other grouping of clusters you would want to perform concurrent operations on as a set.

This is intended for use inside fabrik8, support hub, and other cluster automation tooling.


## Usage

```js
const createClusterInfoClient = require('@npm_wharf/cluster-info-client')

const client = createClusterInfoClient({
    vaultAddress: process.env.VAULT_ADDR || 'http://localhost:8200',
    vaultToken: process.env.VAULT_TOKEN
})

const clusters = await client.listClusters()
```

## Configuration

|name|description|default value|
|---|---|---|
|vaultHost|the address, including protocol, to the vault server|`'http://localhost:8200'`|
|vaultRoleId|a Vault AppRole Role ID with read/write access to cluster data| |
|vaultSecretId|a Vault AppRole Secret ID with read/write access to cluster data| |
|vaultToken|if not using AppRole, a Vault token can be used directly| |
|vaultPrefix|prefix to the Vault data path|`'kv/'`|

## Backend

### Vault

Vault will have:

- a map of Google service account keys at `/credentials/google/${serviceAccountEmail}`
- a map of AWS IAM keys at `/credentials/amazon/${accountName}`
- common provider specific data shared amongst clusters at `/clusters/common/${provider}`
- a `/clusters/all` record whose value is a map of all `clusterSlug`s to their path in Vault, eg `mycluster: '/clusters/production/mycluster'`
- a map of clusters `/clusters/${environment}/${clusterSlug}` containing all information specific to a single cluster.  The record will have
  + A `value` key with a JSON blob of general cluster info, as defined by the schema.
  + A `channels` key with a JSON array of all channels the cluster belongs to
- a `/channels/all` record whose `value` is a JSON array of all channels
- a series of `/channels/${channel}` records whose `value` is a JSON array of `clusterSlug`s, that enumerates all clusters within a given channel

## API

All `client` methods are async and return Promises.

### getCluster(slug)

Gets all information about a single cluster.

```js
const cluster = await client.getCluster('some-cluster')

assert.eql(cluster, {
    name: 'some-cluster',
    environment: 'production',
    props: {
        password: 'hunter2'
    },
    channels: ['production', 'p100']
})
```

### listClusters()

Lists the names of all registered clusters.

```js
const clusters = await client.listClusters()

assert.eql(clusters, [
    'some-cluster',
    'another-cluster'
])
```

### registerCluster(slug, environment, props, [channels])

Registers a single cluster in the system.  `props` go in vault.  You can also optionally list the channels the cluster should be in.  An error will be thrown if any of the channels do not exist.

```js
await client.addCluster('some-cluster', 'staging', {password: 'hunter2'}, ['production'])
```

### updateCluster(slug, newEnvironment, newProps)

Modify an existing cluster, replacing the existing `environment` and/or `props` with new sets. It is best to `getCluster` beforehand.

```js
await client.updateCluster('some-cluster', 'dev', {password: 'hunter3'})
```

### unregisterCluster(slug)

Un-register a cluster with the system.  Channel associations will also be removed.

```js
await client.removeCluster('some-cluster')
```

### listChannels()

List all the registered channels.

```js
const channels = await client.listChannels()
assert.eql(channels, [
    'production',
    'staging',
    'p100',
    'p200',
    'p300'
])
```

### createChannel(channel)

Create a channel, so clusters can be added to it.

```js
await client.createChannel('integration')
```

### removeChannel(channel)

Removes a channel.  Channel must have no clusters associated with it, otherwise an error is thrown.

```js
await client.removeChannel('integration')
```

### addClusterToChannel(slug, channel)

Adds a cluster to a channel.  The cluster and channel must exist, otherwise an error is thrown.

```js
await client.addClusterToChannel('some-cluster', 'production')
```

### removeClusterFromChannel(slug, channel)

Removes a cluster to from a channel.

```js
await client.removeClusterFromChannel('some-cluster', 'integration')
```

### listClustersByChannel(channel)

Lists the names of all clusters within the specified cluster.

```js
const clusters = await listClustersByChannel('production')

assert.eql(clusters, [
    'some-cluster',
    'another-cluster'
])
```

### addServiceAccount(jsonServiceAccountKey)

Store a Service Account private key.  Must be in JSON format.

```js
const key = {
  'type': 'service_account',
  'project_id': 'my-project',
  'private_key_id': 'cccaedf234c3a4de25fce3adf245cfae352d4',
  'private_key': '-----BEGIN PRIVATE KEY-----\n...\n-----END PRIVATE KEY-----\n',
  'client_email': 'my-sa1@my-project.iam.gserviceaccount.com',
  'client_id': '22349780582375098234759',
  'auth_uri': 'https://accounts.google.com/o/oauth2/auth',
  'token_uri': 'https://oauth2.googleapis.com/token',
  'auth_provider_x509_cert_url': 'https://www.googleapis.com/oauth2/v1/certs',
  'client_x509_cert_url': 'https://www.googleapis.com/robot/v1/metadata/x509/my-sa1%40my-project.iam.gserviceaccount.com'
}
await addServiceAccount(key)
```

### getServiceAccount(serviceAccountAddress)

Fetch a Service Account private  key, given a `client_email`

```js
const key = await getServiceAccount('my-sa1@my-project.iam.gserviceaccount.com')
```

### removeServiceAccount(serviceAccountAddress)

Delete a Service Account private  key, given a `client_email`

```js
await removeServiceAccount('my-sa1@my-project.iam.gserviceaccount.com')
```

### listServiceAccounts()

List the addresses of all stored Service Account keys.

```js
const list = await listServiceAccounts()

assert.eql(list, [
    'my-sa1@my-project.iam.gserviceaccount.com',
    'my-sa2@my-project.iam.gserviceaccount.com',
    ...
])
```

### getCommon(provider)

Get common defaults/values for all clusters for a given provider.  (Default: `'GKE'`)

```js
const data = await getCommon(provider)

assert.eql(data, {
  authAccount: 'resource-manager@iam.googleapis.com',
  billingAccount: 123412341234,
  organizationId: 678967896789,
  clusterDefaults: {
    //...
  }
  //...
})
```

### issueCertificate(role, domain, ttl = 5 * 60)

Issue a child certificate using Vault's PKI engine.

```js
const resp = await issueCertificate('support-hub', 'my-cluster.npme.io')

assert.eql(resp, {
  certificate: '---- {PEM string}...',
  expiration: 1559166104,
  issuing_ca: '---- {PEM string}...',
  private_key: '---- {key string}...',
  private_key_type: 'rsa',
  serial_number: '52:be:c0:65:33:fa:6e:aa:02:60:10:be:c5:f7:f6:f4:a1:2c:3c:dc'
})
```
### close()

Closes the client and any associated connections to backends.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.8](https://github.com/npm-wharf/cluster-info-client/compare/v2.0.7...v2.0.8) (2019-07-26)


### Bug Fixes

* ensure cluster exists before adding channels ([793e0b3](https://github.com/npm-wharf/cluster-info-client/commit/793e0b3))



## [2.0.7](https://github.com/npm-wharf/cluster-info-client/compare/v2.0.6...v2.0.7) (2019-07-24)


### Bug Fixes

* don't clobber exisiting channels when registering over an existinc cluster ([477c342](https://github.com/npm-wharf/cluster-info-client/commit/477c342))



## [2.0.6](https://github.com/npm-wharf/cluster-info-client/compare/v2.0.5...v2.0.6) (2019-07-17)


### Bug Fixes

* credentials and masterEndpoint won't be known when initially saving ([97c937e](https://github.com/npm-wharf/cluster-info-client/commit/97c937e))



## [2.0.5](https://github.com/npm-wharf/cluster-info-client/compare/v2.0.4...v2.0.5) (2019-07-15)



## [2.0.4](https://github.com/npm-wharf/cluster-info-client/compare/v2.0.3...v2.0.4) (2019-07-12)


### Bug Fixes

* dont needlessly overwrite data ([0a418c9](https://github.com/npm-wharf/cluster-info-client/commit/0a418c9))



## [2.0.3](https://github.com/npm-wharf/cluster-info-client/compare/v2.0.2...v2.0.3) (2019-07-12)


### Bug Fixes

* auth before ensuring cluster exists ([ac2c411](https://github.com/npm-wharf/cluster-info-client/commit/ac2c411))



## [2.0.2](https://github.com/npm-wharf/cluster-info-client/compare/v2.0.1...v2.0.2) (2019-07-12)


### Bug Fixes

* username is actually user in the schema ([bb5e4b8](https://github.com/npm-wharf/cluster-info-client/commit/bb5e4b8))



## [2.0.1](https://github.com/npm-wharf/cluster-info-client/compare/v2.0.0...v2.0.1) (2019-07-10)


### Bug Fixes

* allow 'dev' as an environment ([a4597d2](https://github.com/npm-wharf/cluster-info-client/commit/a4597d2))



## [1.9.1](https://github.com/npm-wharf/cluster-info-client/compare/v1.9.0...v1.9.1) (2019-07-09)


### Bug Fixes

* **schema:** remove redundant common projectNumber property ([322c498](https://github.com/npm-wharf/cluster-info-client/commit/322c498))



# [1.9.0](https://github.com/npm-wharf/cluster-info-client/compare/v1.8.1...v1.9.0) (2019-06-27)


### Features

* add schema for cluster-info data ([#9](https://github.com/npm-wharf/cluster-info-client/issues/9)) ([1171ef4](https://github.com/npm-wharf/cluster-info-client/commit/1171ef4))



## [1.8.1](https://github.com/npm-wharf/cluster-info-client/compare/v1.8.0...v1.8.1) (2019-06-21)


### Bug Fixes

* workaround for stack overflow bug ([6d660fc](https://github.com/npm-wharf/cluster-info-client/commit/6d660fc))



# [1.8.0](https://github.com/npm-wharf/cluster-info-client/compare/v1.7.1...v1.8.0) (2019-06-05)


### Features

* idempotent adding of service accounts ([db7ac42](https://github.com/npm-wharf/cluster-info-client/commit/db7ac42))



## [1.7.1](https://github.com/npm-wharf/cluster-info-client/compare/v1.7.0...v1.7.1) (2019-06-04)


### Bug Fixes

* **docs:** parameter name ([84b1cf2](https://github.com/npm-wharf/cluster-info-client/commit/84b1cf2))
* await approle before issueing cert ([ac64d9b](https://github.com/npm-wharf/cluster-info-client/commit/ac64d9b))



# [1.7.0](https://github.com/npm-wharf/cluster-info-client/compare/v1.6.0...v1.7.0) (2019-05-31)


### Features

* add issueCertificate ([55ff8ab](https://github.com/npm-wharf/cluster-info-client/commit/55ff8ab))



# [1.6.0](https://github.com/npm-wharf/cluster-info-client/compare/v1.5.0...v1.6.0) (2019-05-04)


### Features

* approle support ([6436aa8](https://github.com/npm-wharf/cluster-info-client/commit/6436aa8))



# [1.5.0](https://github.com/npm-wharf/cluster-info-client/compare/v1.4.2...v1.5.0) (2019-05-01)


### Features

* add listClustersByChannel ([d963404](https://github.com/npm-wharf/cluster-info-client/commit/d963404))



## [1.4.2](https://github.com/npm-wharf/cluster-info-client/compare/v1.4.1...v1.4.2) (2019-04-26)


### Bug Fixes

* dont write data to vault if its the same ([1f6ffa8](https://github.com/npm-wharf/cluster-info-client/commit/1f6ffa8))



## [1.4.1](https://github.com/npm-wharf/cluster-info-client/compare/v1.4.0...v1.4.1) (2019-04-23)



# [1.4.0](https://github.com/npm-wharf/cluster-info-client/compare/v1.3.0...v1.4.0) (2019-04-11)


### Features

* add getCommon method for defaults among providers ([#7](https://github.com/npm-wharf/cluster-info-client/issues/7)) ([604567b](https://github.com/npm-wharf/cluster-info-client/commit/604567b))



# [1.3.0](https://github.com/npm-wharf/cluster-info-client/compare/v1.2.0...v1.3.0) (2019-04-06)


### Features

* make vault-prefix configurable ([#5](https://github.com/npm-wharf/cluster-info-client/issues/5)) ([77dd1e9](https://github.com/npm-wharf/cluster-info-client/commit/77dd1e9))



# [1.2.0](https://github.com/npm-wharf/cluster-info-client/compare/v1.1.0...v1.2.0) (2019-04-04)


### Features

* initial implementation ([#2](https://github.com/npm-wharf/cluster-info-client/issues/2)) ([4482bf4](https://github.com/npm-wharf/cluster-info-client/commit/4482bf4))
* store secrets in vault ([#3](https://github.com/npm-wharf/cluster-info-client/issues/3)) ([02805a9](https://github.com/npm-wharf/cluster-info-client/commit/02805a9))
* Store service accounts in vault ([#4](https://github.com/npm-wharf/cluster-info-client/issues/4)) ([d2f4ce2](https://github.com/npm-wharf/cluster-info-client/commit/d2f4ce2))



# 1.1.0 (2019-03-22)


### Features

* initial implementation ([a652304](https://github.com/npm-wharf/cluster-info-client/commit/a652304))


index.js
'use strict'

const createVault = require('node-vault')
const rpn = require('request-promise-native')
const equal = require('fast-deep-equal')
const SA_PATH = 'credentials/google/'

module.exports = function createClient (options = {}) {
  const {
    vaultHost = 'http://localhost:8200',
    vaultToken = process.env.VAULT_TOKEN,
    vaultRoleId = process.env.VAULT_ROLE_ID,
    vaultSecretId = process.env.VAULT_SECRET_ID,
    vaultPrefix = process.env.VAULT_SECRET_PREFIX || 'kv/'
  } = options

  let vault = createVault({
    endpoint: vaultHost,
    token: vaultToken,
    // workaround for https://github.com/kr1sp1n/node-vault/issues/80
    'request-promise': rpn
  })

  let vaultAuth

  if (vaultToken) {
    vaultAuth = Promise.resolve()
  } else if (vaultRoleId && vaultSecretId) {
    vaultAuth = vault.approleLogin({
      role_id: vaultRoleId,
      secret_id: vaultSecretId
    }).then(result => {
      const { auth } = result
      vault = createVault({
        endpoint: vaultHost,
        token: auth.client_token,
        'request-promise': rpn
      })
    })
  }

  return {
    createChannel,
    deleteChannel,
    listChannels,

    registerCluster,
    updateCluster,
    unregisterCluster,
    listClusters,
    getCluster,

    addClusterToChannel,
    removeClusterFromChannel,
    listClustersByChannel,

    addServiceAccount,
    getServiceAccount,
    removeServiceAccount,
    listServiceAccounts,

    getCommon,

    issueCertificate,

    close
  }

  async function registerCluster (slug, environment, secretProps, channels = []) {
    if (typeof environment !== 'string') throw new Error('`environment` must be a string')
    await vaultAuth
    await _ensureChannelsExist(channels)

    const props = {
      environment,
      value: secretProps
    }

    await _saveClusterData(props, slug, environment)

    const { value = {} } = await _readVault(_allClustersPath())
    const path = _secretPath(slug, environment)
    if (value[slug] !== path) {
      value[slug] = path
      await vault.write(_allClustersPath(), { data: { value: JSON.stringify(value, null, 2) } })
    }

    for (const chan of channels) {
      await addClusterToChannel(slug, chan)
    }
  }

  async function updateCluster (slug, environment, secretProps) {
    await vaultAuth
    await _ensureClusterExists(slug)

    const props = {
      value: secretProps,
      environment
    }

    await _saveClusterData(props, slug, environment)
  }

  async function unregisterCluster (slug) {
    await vaultAuth
    await _ensureClusterExists(slug)

    const { channels, environment } = await getCluster(slug)

    // un-associate the cluster from channels
    await Promise.all(channels.map(channel => _removeClusterFromChannel(slug, channel)))

    const { value: allClusters } = await _readVault(_allClustersPath())
    delete allClusters[slug]
    await vault.write(_allClustersPath(), { data: { value: JSON.stringify(allClusters, null, 2) } })

    await _deleteSecret(slug, environment)
  }

  async function listClusters () {
    const { value } = await _readVault(_allClustersPath())
    return Object.keys(value).sort()
  }

  async function getCluster (slug) {
    await _ensureClusterExists(slug)
    const { value: clusterMap } = await _readVault(_allClustersPath())
    const environment = /clusters\/([^/]+)\//.exec(clusterMap[slug])[1]
    return _getSecret(slug, environment)
  }

  async function createChannel (channel) {
    if (channel === 'all') return

    const list = await listChannels()
    if (list.includes(channel)) return
    list.push(channel)
    list.sort()

    await vault.write(_channelPath('all'), {
      data: { value: JSON.stringify(list, null, 2) }
    })

    await vault.write(_channelPath(channel), {
      data: { value: '[]' }
    })
  }

  async function deleteChannel (channel) {
    const list = await listChannels()
    const newList = list.filter(c => c !== channel)

    await vault.write(`${vaultPrefix}data/channels/all`, {
      data: { value: JSON.stringify(newList, null, 2) }
    })

    await vault.delete(`${vaultPrefix}data/channels/${channel}`)
  }

  async function listChannels () {
    await vaultAuth
    const { value = [] } = await _readVault(`${vaultPrefix}data/channels/all`)
    return value
  }

  async function addClusterToChannel (slug, channel) {
    await vaultAuth
    const [ cluster ] = await Promise.all([
      getCluster(slug),
      _ensureChannelsExist([channel])
    ])

    let { channels = [], environment } = cluster

    if (!channels.includes(channel)) {
      channels.push(channel)
      channels.sort()
    }

    await Promise.all([
      _addClusterToChannel(slug, channel),
      _saveClusterData({ channels }, slug, environment)
    ])
  }

  async function removeClusterFromChannel (slug, channel) {
    await vaultAuth
    const [ cluster ] = await Promise.all([
      getCluster(slug),
      _ensureChannelsExist([channel])
    ])

    let { channels, environment } = cluster

    channels = channels.filter(chan => chan !== channel)

    await Promise.all([
      _removeClusterFromChannel(slug, channel),
      _saveClusterData({ channels }, slug, environment)
    ])
  }

  async function listClustersByChannel (channel) {
    await vaultAuth
    const { value } = await _readVault(_channelPath(channel))
    return value
  }

  async function addServiceAccount (jsonServiceAccountKey) {
    const addr = jsonServiceAccountKey['client_email']
    if (!addr) throw new Error('service account key must have a `client_email` property')
    await vaultAuth
    const exisiting = await getServiceAccount(addr)

    if (equal(exisiting, jsonServiceAccountKey)) return

    await vault.write(_secretSAPath(addr), { data: { value: JSON.stringify(jsonServiceAccountKey, null, 2) } })
  }

  async function getServiceAccount (serviceAccountAddress) {
    const resp = await _readVault(_secretSAPath(serviceAccountAddress))
    return resp.value
  }

  async function removeServiceAccount (serviceAccountAddress) {
    await vaultAuth
    await vault.delete(_secretSAPath(serviceAccountAddress))
  }

  async function listServiceAccounts () {
    await vaultAuth
    const resp = await vault.list(`${vaultPrefix}metadata/${SA_PATH}`)
    return resp.data.keys.sort()
  }

  async function getCommon (provider = 'GKE') {
    return (await _readVault(`${vaultPrefix}data/clusters/common/${provider.toLowerCase()}`)).value
  }

  async function issueCertificate (role, domain, ttl = 5 * 60) {
    await vaultAuth
    const resp = await vault.write(`pki/issue/${role}`, { common_name: domain, ttl: ttl })
    return resp.data
  }

  function close () {
  }

  async function _ensureClusterExists (slug) {
    await vaultAuth
    const resp = await vault.read(`${vaultPrefix}data/clusters/all`)
    const map = JSON.parse(resp.data.data.value)
    if (!map[slug]) throw new Error(`cluster '${slug}' does not exist`)
  }

  async function _ensureChannelsExist (channels) {
    const list = await listChannels()

    channels.forEach(channel => {
      if (!list.includes(channel)) {
        throw new Error(`channel '${channel}' must exist`)
      }
    })
  }

  async function _addClusterToChannel (slug, channel) {
    const { value: clusters } = await _readVault(_channelPath(channel))

    if (clusters.includes(slug)) return

    clusters.push(slug)
    clusters.sort()
    await vault.write(_channelPath(channel), { data: { value: JSON.stringify(clusters, null, 2) } })
  }

  async function _removeClusterFromChannel (slug, channel) {
    const { value: clusters } = await _readVault(_channelPath(channel))

    const newClusters = clusters.filter(cl => cl !== slug)

    if (newClusters.length === clusters.length) return

    await vault.write(_channelPath(channel), { data: { value: JSON.stringify(newClusters, null, 2) } })
  }

  async function _saveClusterData (props, slug, environment) {
    const previous = await _readVault(_secretPath(slug, environment))

    // only update if a key has changed
    const needsUpdate = Object.keys(props).some(key => {
      if ((typeof props[key] === 'string') && previous[key] !== props[key]) return true
      if (!equal(previous[key], props[key])) return true
    })

    if (!needsUpdate) return

    await vault.write(_secretPath(slug, environment), {
      data: mapValues({ ...previous, ...props }, val => {
        if (typeof val === 'string') return val
        return JSON.stringify(val, null, 2)
      })
    })
  }

  async function _getSecret (slug, environment) {
    return _readVault(_secretPath(slug, environment))
  }

  async function _readVault (path) {
    await vaultAuth
    try {
      var resp = await vault.read(path)
    } catch (err) {
      // istanbul ignore next
      if (err.response && err.response.statusCode === 404) return {}
      // istanbul ignore next
      throw err
    }
    const result = mapValues(resp.data.data, val => {
      let parsed = val
      try {
        parsed = JSON.parse(val)
      } catch (e) {}
      return parsed
    })
    return result
  }

  async function _deleteSecret (name, environment) {
    await vaultAuth
    await vault.delete(_secretPath(name, environment))
  }

  function _secretPath (name, environment) {
    return `${vaultPrefix}data/clusters/${environment}/${name}`
  }

  function _allClustersPath () {
    return `${vaultPrefix}data/clusters/all`
  }

  function _secretSAPath (addr) {
    return `${vaultPrefix}data/${SA_PATH}${addr}`
  }

  function _channelPath (channel) {
    return `${vaultPrefix}data/channels/${channel}`
  }
}

function mapValues (obj, fn) {
  const newObj = {}
  Object.keys(obj).forEach(key => {
    newObj[key] = fn(obj[key])
  })
  return newObj
}
